### PR TITLE
LibraryPanels: Add folder checks and permissions

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Relates #31305

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/services/librarypanels/guard.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+import (
+	"fmt"
+
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+)
+
+func isGeneralFolder(folderID int64) bool {
+	return folderID == 0
```
Comment (hugohaggmark) [inline pkg/services/librarypanels/guard.go:11]: Is this a correct assumption?

File: pkg/services/librarypanels/guard.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+func doesUserHaveEditorPermissionsOnFolder(session *sqlstore.DBSession, user *models.SignedInUser, folderID int64) (bool, error) {
+	if isGeneralFolder(folderID) && user.HasRole(models.ROLE_EDITOR) {
+		return true, nil
+	}
+
+	if isGeneralFolder(folderID) && user.HasRole(models.ROLE_VIEWER) {
+		return false, models.ErrFolderAccessDenied
+	}
```
Comment (hugohaggmark) [inline pkg/services/librarypanels/guard.go:43]: Are these correct assumptions?

File: pkg/services/librarypanels/database.go [hunk 3]
```diff
@@ -399,6 +405,30 @@ func (lps *LibraryPanelService) getLibraryPanelsForDashboardID(c *models.ReqCont
 	return libraryPanelMap, err
 }
 
+func handleFolderIDPatches(session *sqlstore.DBSession, panelToPatch *LibraryPanel, fromFolderID int64, toFolderID int64, user *models.SignedInUser) error {
+	// FolderID was not provided in the PATCH request
+	if toFolderID == -1 {
+		panelToPatch.FolderID = fromFolderID
```
Comment (aknuds1) [inline pkg/services/librarypanels/database.go:411]: I'm wondering if the following makes for less complex/more straightforward code?
```suggestion
		toFolderID = fromFolderID
```

**Review by torkelo — APPROVED** [review]
Looks great, only issue I can see is with doesUserHaveEditorPermissionsOnFolder and if that should call out to FolderService and guardian

File: pkg/services/librarypanels/guard.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	if count == 0 {
+		return false, models.ErrFolderAccessDenied
+	}
+	if count > 1 {
+		return false, fmt.Errorf("found %d folders, while expecting at most one", count)
+	}
+
+	return true, nil
+}
```
[the remark this answers is not in the document]
Comment (hugohaggmark) [reply pkg/services/librarypanels/guard.go:63]: Ok, excellent suggestions I'll have a look.

File: pkg/services/librarypanels/database.go (hunk 3 above)
Comment (hugohaggmark) [reply pkg/services/librarypanels/database.go:411]: I think this makes sense, but I need to update panelToPatch too.

File: pkg/services/librarypanels/guard.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
+package librarypanels
+
+import (
+	"fmt"
+
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+)
+
+func isGeneralFolder(folderID int64) bool {
```
[the remark this answers is not in the document]
Comment (hugohaggmark) [reply pkg/services/librarypanels/guard.go:10]: I only found a lot of `FolderID > 0` in the code, so not sure we could share this.

File: pkg/services/librarypanels/guard.go (hunk 1 above)
Comment (torkelo) [reply pkg/services/librarypanels/guard.go:10]: yes, looks like it :)

File: pkg/services/librarypanels/guard.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (hugohaggmark) [reply pkg/services/librarypanels/guard.go]: Fixed

**Review by torkelo — APPROVED** [review]
Looks good, great job!

**Pull request merged**